nav2_behaviors/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,203 @@
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::joyCallback,
+      this, std::placeholders::_1));
+}
+
+Status AssistedTeleop::onRun(const std::shared_ptr<const AssistedTeleopAction::Goal> command)
+{
+  command_time_allowance_ = command->time_allowance;
+  end_time_ = steady_clock_.now() + command_time_allowance_;
+  return Status::SUCCEEDED;
+}
+
+void AssistedTeleop::onActionCompletion()
+{
+  teleop_twist_ = geometry_msgs::msg::Twist();
+  joy_ = sensor_msgs::msg::Joy();
+}
+
+Status AssistedTeleop::onCycleUpdate()
+{
+  feedback_->current_teleop_duration = elasped_time_;
+  action_server_->publish_feedback(feedback_);
+
+  rclcpp::Duration time_remaining = end_time_ - steady_clock_.now();
+  if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
+    stopRobot();
+    RCLCPP_WARN_STREAM(
+      logger_,
+      "Exceeded time allowance before reaching the " << behavior_name_.c_str() <<
+        "goal - Exiting " << behavior_name_.c_str());
+    return Status::FAILED;
+  }
+
+  // user states that teleop was successful
+  if (joy_.buttons.size() > 0 && joy_.buttons[0] > 0) {
+    stopRobot();
+    return Status::SUCCEEDED;
+  }
+
+  // user states that teleop failed
+  if (joy_.buttons.size() > 1 && joy_.buttons[1] > 0) {
+    stopRobot();
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::Twist cmd_vel;
+  cmd_vel = computeVelocity(teleop_twist_);
+  vel_pub_->publish(std::move(cmd_vel));
+
+  return Status::RUNNING;
+}
+
+geometry_msgs::msg::Twist AssistedTeleop::computeVelocity(geometry_msgs::msg::Twist & twist)
+{
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR_STREAM(
+      logger_,
+      "Current robot pose is not available for " <<
+        behavior_name_.c_str());
+    return geometry_msgs::msg::Twist();
+  }
+  geometry_msgs::msg::Pose2D projected_pose;
+  projected_pose.x = current_pose.pose.position.x;
+  projected_pose.y = current_pose.pose.position.y;
+  projected_pose.theta = tf2::getYaw(current_pose.pose.orientation);
+
+  geometry_msgs::msg::Twist scaled_twist = twist;
+  for (double time = simulation_time_step_; time < projection_time_;
+    time += simulation_time_step_)
+  {
+    projected_pose = projectPose(projected_pose, twist, simulation_time_step_);
+
+    if (!collision_checker_->isCollisionFree(projected_pose)) {
+      if (time == simulation_time_step_) {
+        RCLCPP_WARN_STREAM_THROTTLE(
+          logger_,
+          steady_clock_,
+          500,
+          behavior_name_.c_str() << " collided on first time step, setting velocity to zero");
+        scaled_twist.linear.x = 0.0f;
+        scaled_twist.linear.y = 0.0f;
+        scaled_twist.angular.z = 0.0f;
+        break;
+      } else {
+        RCLCPP_WARN_STREAM_THROTTLE(
```

Comment:
Debug log

Reply:
The point of this is to stop collisions so we don't need to blast them that this behavior is doing its job :-)